cirq-core/cirq/value/classical_data_test.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+
+import cirq
+
+mkey_m = cirq.MeasurementKey('m')
+mkey_c = cirq.MeasurementKey('c')
+two_qubits = tuple(cirq.LineQubit.range(2))
+
+
+def test_init():
+    cd = cirq.ClassicalDataDictionaryStore()
+    assert cd.measurements == {}
+    assert cd.keys() == ()
+    assert cd.measured_qubits == {}
+    assert cd.channel_measurements == {}
+    assert cd.measurement_types == {}
+    cd = cirq.ClassicalDataDictionaryStore(
+        _measurements={mkey_m: (0, 1)},
+        _measured_qubits={mkey_m: two_qubits},
+        _channel_measurements={mkey_c: 3},
+    )
+    assert cd.measurements == {mkey_m: (0, 1)}
+    assert cd.keys() == (mkey_m, mkey_c)
+    assert cd.measured_qubits == {mkey_m: two_qubits}
+    assert cd.channel_measurements == {mkey_c: 3}
+    assert cd.measurement_types == {
+        mkey_m: cirq.MeasurementType.MEASUREMENT,
+        mkey_c: cirq.MeasurementType.CHANNEL,
+    }
+
+
+def test_record_measurement():
+    cd = cirq.ClassicalDataDictionaryStore()
+    cd.record_measurement(mkey_m, (0, 1), two_qubits)
+    assert cd.measurements == {mkey_m: (0, 1)}
+    assert cd.keys() == (mkey_m,)
+    assert cd.measured_qubits == {mkey_m: two_qubits}
+
+
+def test_record_measurement_errors():
+    cd = cirq.ClassicalDataDictionaryStore()
+    with pytest.raises(ValueError, match='3 measurements but 2 qubits'):
+        cd.record_measurement(mkey_m, (0, 1, 2), two_qubits)
+    cd.record_measurement(mkey_m, (0, 1), two_qubits)
+    with pytest.raises(ValueError, match='Measurement already logged to key m'):
+        cd.record_measurement(mkey_m, (0, 1), two_qubits)
+
+
+def test_record_channel_measurement():
+    cd = cirq.ClassicalDataDictionaryStore()
+    cd.record_channel_measurement(mkey_m, 1)
+    assert cd.channel_measurements == {mkey_m: 1}
+    assert cd.keys() == (mkey_m,)
+
+
+def test_record_channel_measurement_errors():
```

Comment:
Add tests for cross-type errors (i.e. overwriting "normal" measurement with channel measurement and vice-versa)